include/deal.II/base/thread_local_storage.h

```diff
@@ -193,40 +179,59 @@ namespace Threads
   // ----------------- inline and template functions --------------------------
 
   template <typename T>
-  inline ThreadLocalStorage<T>::ThreadLocalStorage(const T &t)
-    : data(t)
+  inline ThreadLocalStorage<T>::ThreadLocalStorage()
   {}
 
 
   template <typename T>
-  inline ThreadLocalStorage<T>::ThreadLocalStorage(
-    const ThreadLocalStorage<T> &t)
-    : data(t)
-  {}
+  inline ThreadLocalStorage<T>::ThreadLocalStorage(const T &t)
+  {
+    exemplar = std::make_shared<T>(t);
+  }
 
 
   template <typename T>
   inline T &
-  ThreadLocalStorage<T>::get()
+  ThreadLocalStorage<T>::get(bool &exists)
   {
-#    ifdef DEAL_II_WITH_THREADS
-    return data.local();
-#    else
-    return data;
-#    endif
+    const std::thread::id my_id = std::this_thread::get_id();
+
+    auto it = data.find(my_id);
+    if (it != data.end())
+      {
+        exists = true;
+        return it->second;
+      }
+
+    exists = false;
+
+    std::lock_guard lock(insertion_mutex);
+
+    // FIXME: Port to C++11
+
+    if constexpr (std::is_trivially_copyable<T>::value)
+      if (exemplar)
+        {
+          it = data.insert(std::make_pair(my_id, *exemplar)).first;
```

Comment:
I think you can use `emplace` and then all you need is that the object is copy constructible.